### PR TITLE
ambient: stash hbone peer principal in endpoint metadata

### PR DESCRIPTION
**Pull request by keithmattix**
**Please provide a description of this PR:**
If envoyproxy/envoy#33857 merges, we'll be able to use `io.istio.upstream_peer_principal` to grab the destination principal in telemetry

### CODE REVIEW

File: pilot/pkg/xds/filters/filters.go [hunk 1]
```diff
@@ -333,7 +333,7 @@ var (
 								Format: &core.SubstitutionFormatString_TextFormatSource{
 									TextFormatSource: &core.DataSource{
 										Specifier: &core.DataSource_InlineString{
-											InlineString: "%DOWNSTREAM_PEER_URI_SAN%",
+											InlineString: "%UPSTREAM_PEER_URI_SAN%",
```
Comment (howardjohn) [inline pilot/pkg/xds/filters/filters.go:336]: io.istio.peer_principal is the thing we are asserting rbac against, which should be downstream. maybe we need a new key for upstream for stats/logging

Comment (keithmattix) [reply pilot/pkg/xds/filters/filters.go:336]: Ah didn't realize we use it for rbac. I could do _upstream_peer_principal?

Comment (howardjohn) [reply pilot/pkg/xds/filters/filters.go:336]: @kyessenov may have a good opinion here since they did most of this metadata stuff?

Comment (kyessenov) [reply pilot/pkg/xds/filters/filters.go:336]: Not sure I understand the context of this.

**Comment by keithmattix** [comment]
This is ready for review now; we decided to stash the identity in metadata due to complexity

File: pilot/pkg/xds/filters/filters.go (hunk 1 above)
Comment (keithmattix) [reply pilot/pkg/xds/filters/filters.go]: Did something different

File: pilot/pkg/networking/core/cluster_waypoint.go [hunk 2]
```diff
@@ -57,9 +58,32 @@ func buildInternalUpstreamCluster(name string, internalListener string) *cluster
 	}
 }
 
+func buildEncapInternalUpstreamCluster(name string, internalListener string) *cluster.Cluster {
```
Comment (howardjohn) [inline pilot/pkg/networking/core/cluster_waypoint.go:61]: This seems identical to buildInternalUpstreamCluster. is it?

File: pilot/pkg/xds/endpoints/endpoint_builder.go [hunk 3]
```diff
@@ -667,6 +667,10 @@ func buildEnvoyLbEndpoint(b *EndpointBuilder, e *model.IstioEndpoint, mtlsEnable
 		ep.HostIdentifier = &endpoint.LbEndpoint_Endpoint{Endpoint: &endpoint.Endpoint{
 			Address: util.BuildInternalAddressWithIdentifier(connectOriginate, net.JoinHostPort(address, strconv.Itoa(port))),
 		}}
+		peerPrincipal, _ := structpb.NewStruct(map[string]any{
```
Comment (howardjohn) [inline pilot/pkg/xds/endpoints/endpoint_builder.go:670]: Do we need this for everything doing hbone or just waypoints?

Like does sidecar have the issue today? Seems like probably yes?

**Comment by kyessenov** [comment]
I really don't understand the overall design. We mostly stopped using `io.istio.upstream_peer_principal` or other custom Istio states for standard sidecar xDS, so this is going back to that solution on the upstream? You already have WDS in waypoints - why do you need to change xDS if you can look up all telemetry there?

File: pilot/pkg/networking/core/cluster_waypoint.go (hunk 2 above)
Comment (keithmattix) [reply pilot/pkg/networking/core/cluster_waypoint.go:61]: Yeah this is from a former iteration where I had a custom filter. I'll clean this up once I get this all working locally

File: pilot/pkg/xds/endpoints/endpoint_builder.go (hunk 3 above)
Comment (keithmattix) [reply pilot/pkg/xds/endpoints/endpoint_builder.go:670]: Yeah I would think so; the simplest implementation seemed to be just always adding the identity in endpoint metadata

**Comment by keithmattix** [comment]
> I really don't understand the overall design. We mostly stopped using `io.istio.upstream_peer_principal` or other custom Istio states for standard sidecar xDS, so this is going back to that solution on the upstream? You already have WDS in waypoints - why do you need to change xDS if you can look up all telemetry there?

I didn't realize waypoints were requesting info via WDS; which filter is doing that? Metadata exchange via discovery?

**Comment by kyessenov** [comment]
@keithmattix It's done in the metadata exchange extension, but the code is just a lookup so you can do the same directly in the stats filter. This will hurt our ability to upstream the stats filter because WDS is non-standard.

**Comment by keithmattix** [comment]
@kyessenov I found it, but the implementation required me to edit the flatbuffer so there's a spot for the identity. It's probably way too heavy for what we need, but I'll push it up once I confirm it works for you to take a look

**Comment by kyessenov** [comment]
@keithmattix I'd prefer keeping ambient stuff to ambient code, so if we can avoid changing EDS for ambient, that's better. Adding another field to flatbuffer is not a problem, we have to refactor it anyways to support Otel telemetry.

**Comment by keithmattix** [comment]
> @keithmattix I'd prefer keeping ambient stuff to ambient code, so if we can avoid changing EDS for ambient, that's better. Adding another field to flatbuffer is not a problem, we have to refactor it anyways to support Otel telemetry.

Ack - I just pushed up my stash. Serializing to the WorkloadObject is returning an empty string for some reason, and when I try to access the flat buffer directly, it looks corrupted.

**Comment by keithmattix** [comment]
Converting back to draft as I'm still testing and trying to get it to work e2e

**Comment by costinm** [comment]
Agree with Kuat, WDS is the right approach. We may need to iterate on what
is exposed - but identity is not controversial.

I think sidecars and regular gateways should use this as well - not only
ambient.
We are discussing the Istio sidecar and gateway behavior if mTLS is off
(for any reasons) - getting peer info from
MDS is the best option.

**Comment by howardjohn** [comment]
WDS works well for ztunnel because its the source of truth for things. Its IMO a poor choice when its decoupled from the rest of the routing flow (reminds me of authz and HTTPRoute :slightly_smiling_face: ).

For example, what happens when the IP we send to is the east-west GW IP? We want metadata for the underlying destination, not the EW GW.

This information is already in EDS, we already put 4 pieces of metadata there -- and it correctly handles these cases. Why put a 5th in a different location?

**Comment by kyessenov** [comment]
> This information is already in EDS, we already put 4 pieces of metadata there -- and it correctly handles these cases. Why put a 5th in a different location?

Because we know that was not a good solution. Not every endpoint is in a cluster to start with, you have to support pass-through cases just the same way. There's no reason why WDS/MDS can't support gateway IPs.

**Comment by howardjohn** [comment]
> Because we know that was not a good solution. Not every endpoint is in a cluster to start with, you have to support pass-through cases just the same way.

Ok but we don't do mTLS for those either, so the identity information doesn't matter. 

Having a disjoint source of information isn't really acceptable on the authorization path

**Comment by kyessenov** [comment]
> > Because we know that was not a good solution. Not every endpoint is in a cluster to start with, you have to support pass-through cases just the same way.
> 
> Ok but we don't do mTLS for those either, so the identity information doesn't matter.
> 
> Having a disjoint source of information isn't really acceptable on the authorization path

Waypoint absolutely relies on WDS for all source telemetry right now. We can just as well rely on it for destination telemetry, instead of trying to make it a hybrid of EDS on the destination side. I would agree if we didn't use WDS already, but it's heavily used at this point on Waypoints.

**Comment by howardjohn** [comment]
TBH I am a bit confused by the position here. In the past, hadn't you advocated for deriving more metadata from trusted attributes? I.e. get the namespace, identity, etc from the TLS handshake. Using WDS seems further from that. What am I missing?

**Comment by kyessenov** [comment]
@howardjohn Yes, when Envoy performs L4 functions as in the sidecar we should use the trusted attributes. But with Ambient, L4 is intentionally walled from L7 so L7 ultimately has to trust something else to provide those attributes. Whether you trust ztunnel on the destination side, Cilium/kernel, or istiod, doesn't really matter - you simply look up identity wherever you can.

Costin points out to a valid model of a sidecar also dropping L4 functions, where we'd need to do this lookup as well, but that's a separate topic.

**Comment by howardjohn** [comment]
Even in the case where you have ztunnel, the model we (or, I) am pushing for is to have an exchange of information between these two layers. Blindly trusting is not a good model IMO. And we are blurring the layers by needing to look up the information out of band instead of just asking the layer.

**Comment by kyessenov** [comment]
> Even in the case where you have ztunnel, the model we (or, I) am pushing for is to have an exchange of information between these two layers. Blindly trusting is not a good model IMO. And we are blurring the layers by needing to look up the information out of band instead of just asking the layer.

I don't think it's a problem. istiod is just a skip-level manager that you can ask instead of going to the direct manager :) There's a general consistency problem that can be improved if you do it on the wire - you can absolutely do the same with an internal protocol between HCM and tcp_proxy instead of raw TCP that read/write principals. But that requires custom filters for a relatively minor feature right now.

**Comment by howardjohn** [comment]
I agree communication between the layers is more work and maybe not worth it, but I don't see how EDS vs WDS is more or less work. Both seem to be just plugging 1 new field into existing mechanisms. Then we will want to use that in SAN match, but probably that work is similar across both methods

**Comment by kyessenov** [comment]
EDS is not meant for policy - it's a load assignment, not even config. Putting a policy for SAN into EDS feels like an abuse of xDS data model.

**Comment by howardjohn** [comment]
Where else would you put per-endpoint policies? If we are talking about abusing XDS data model that does not include WDS either

**Comment by howardjohn** [comment]
The intent (from me, maybe not Keith -- I will let him say) is to use this for telemetry and authz. Our current SAN match is `trust-domain/*` which is obviously unacceptable long term.

> At least with the current Istio mode, authz should only be based on the peer certificate.

I am talking about how to verify the cert

**Comment by costinm** [comment]
On Thu, May 2, 2024 at 5:07 PM John Howard ***@***.***> wrote:

> TBH I am a bit confused by the position here. In the past, hadn't you
> advocated for deriving more metadata from trusted attributes? I.e. get the
> namespace, identity, etc from the TLS handshake. Using WDS seems further
> from that. What am I missing?
>
If we can get signed info - from TLS peer cert, or some JWT - that's
absolutely best.

If we can't get it from a peer cert - the next best source of trust is WDS.
Istiod is also the one handling secure naming
and all ztunnel and sidecar trust decisions are based on info about
captured IP address and the same source of data
as WDS. Not as good as a signed cert or JWT - but still as secure as the
rest of Istio.

Least secure is the header (TCP or http) that peer is sending. Even for
telemetry it's pretty bad, but we have a history.

**Comment by costinm** [comment]
Ok - I am confused... Is this in context of the previous WG meeting and getting principal for telemetry ? Or for some authentication feature that I'm not aware of ( normally CDS has secure naming ) ? 

John - I don't mind if ztunnel can pass the data ( PROXY or other mean)- but ztunnel gets it from Istiod as well. 
Waypoint, sidecar or ztunnel trust Istiod to provide the correct info about an IP - including its service account - 
both for cluster IPs ( secure naming ) and endpoints, and I don't see why a lookup directly with WDS would be
less trusted then getting the data from ztunnel ( less efficient - yes, that's clear)

**Comment by costinm** [comment]
> The intent (from me, maybe not Keith -- I will let him say) is to use this for telemetry and authz. Our current SAN match is `trust-domain/*` which is obviously unacceptable long term.

I would suggest not mixing (again) telemetry and authz. They have very different requirements, and I agree current SAN match is really bad and needs to be fixed. 

> 
> > At least with the current Istio mode, authz should only be based on the peer certificate.
> 
> I am talking about how to verify the cert

I have been thinking about telemetry - and not sure I have full context on the authz issue. 
It can't be about waypoint verifying the client cert - that would be covered by authz policy matches, and 
I assume HA-PROXY discussion for sandwitches. 
Is it about 'secure naming' for waypoints ? That would be covered by CDS for services - and there is no EDS 
for pods that are not in a service. 

Keith - can you add some comments to the PR description to make the use case(s) more clear, and if it is 
security related - it would really, really help to have a doc that can be security reviewed by people who
focus on this area.

**Comment by howardjohn** [comment]
Yes it's securing naming in CDS doesn't work in waypoints due to the
internal listener hop.

If we are already giving a per-endpoint identity in don't see why we would
then aggregate that into a cluster match

**Comment by costinm** [comment]
> ~medium term we definitely need to validate the SAN of the upstream endpoint, otherwise the telemetry on the wire could be false without us knowing. The immediate fix is to try and make sure telemetry is fixed, but I don't see why we shouldn't do SAN validation as well

We should certainly do SAN validation - but not confuse the security and trust model used in authz with the one used 
for telemetry. If we end up with trusted telemetry data - it's great, but not all telemetry can be signed and authenticated
and we would get (again) distracted.  Istio telemetry has not been trust-worthy - and perhaps we can live few more releases with this, since using WDS for telemetry data is far more secure than the current headers and metadata exchange.

I still don't know what 'upstream endpoint' you're validating and where. For clusters - I assume we still use 
CDS and normal process. When the upstream is a pod without CDS - there is no EDS either, so not clear how adding
peer info in EDS would help. To be honest I don't really understand the security model for 'to workload' very
well and I don't think the docs we have are detailed enough ( but I'm not an expert in security - hope someone
does understand the actual details in this area ). But unless something dramatic changed in EDS - a pod that 
is not part of a service should not be there.

**Comment by costinm** [comment]
> Yes it's securing naming in CDS doesn't work in waypoints due to the internal listener hop. If we are already giving a per-endpoint identity in don't see why we would then aggregate that into a cluster match

So this is a HTTP request for "example.ns" that hits waypoint, gets routed to "example-v1" cluster - which has a bunch of 
endpoints. What does the internal hop do ? If Envoy is initiating mTLS - it should have all the info. If it's sandwitch - ztunnel should have the peer info, based on the pod IP. 

Why we aggregate all pod identities in the cluster is a controversial issue, but the current approved design for
secure naming and 'client validating server identity' is what we have. Not perfect - but a change in security
model requires very serious review. Regardless - not aggregating doesn't do much - we still rely on what 
the pod happens to run as instead of what user intends ( I.e. "only example service account can be used for the example service, if any pod with wrong label happens to be deployed - reject" ). At least that's the controversial part
in the secure naming AFAIK - and I believe the Gateway design as I understand it is moving into explicit user
intent instead of aggregation/guessing.

If the request is NOT for a cluster/service - but directly to an endpoint ( workload to workload ) - that's a whole different
discussion.

**Comment by costinm** [comment]
> > @keithmattix I'd prefer keeping ambient stuff to ambient code, so if we can avoid changing EDS for ambient, that's better. Adding another field to flatbuffer is not a problem, we have to refactor it anyways to support Otel telemetry.
> 
> Ack - I just pushed up my stash. Serializing to the WorkloadMetadataObject is returning an empty string for some reason, and when I try to access the flat buffer directly, it looks corrupted. I've never worked with flat buffers, so any advice would be appreciated

I haven't seen the code - how did we end up with flatbuffers ? Didn't we have enough protobuffers and Json and yaml ? Is it something required by envoy ? 

Nothing wrong with flatbuffers - I like them - but seems one more complexity.

**Comment by kyessenov** [comment]
> Nothing wrong with flatbuffers - I like them - but seems one more complexity.

We had flatbuffers since version 1.5 or so :D They were forced by Wasm, there's no good reason to use them now, so it just remains a tech debt.

**Comment by keithmattix** [comment]
Closing because I was able to get this working with WDS

**Pull request closed without merging**